pyscriptjs/src/components/pyscript.ts

```diff
@@ -10,61 +10,129 @@ import { InterpreterClient } from '../interpreter_client';
+        }
+    };
+
+    /**
+     * Given a generic DOM Element, tries to fetch the 'src' attribute, if present.
+     * It either throws an error if the 'src' can't be fetched or it returns a fallback
+     * content as source.
+     */
+    const fetchSource = async (tag: Element, fallback: () => string): Promise<string> => {
+        if (tag.hasAttribute('src')) {
+            try {
+                const response = await robustFetch(tag.getAttribute('src'));
+                return await response.text();
+            } catch (err) {
+                const e = err as Error;
+                _createAlertBanner(e.message);
+                throw e;
+            }
+        }
+        return fallback();
+    };
+
     class PyScript extends HTMLElement {
         srcCode: string;
         stdout_manager: Stdio | null;
         stderr_manager: Stdio | null;
+        _fetchSourceFallback = () => htmlDecode(this.srcCode);
 
         async connectedCallback() {
-            /**
-             * Since connectedCallback is async, multiple py-script tags can be executed in
-             * an order which is not particularly sequential. The locking mechanism here ensures
-             * a sequential execution of multiple py-script tags present in one page.
-             *
-             * Concurrent access to the multiple py-script tags is thus avoided.
-             */
-            app.incrementPendingTags();
-            let releaseLock: () => void;
-            try {
-                releaseLock = await app.tagExecutionLock();
-                ensureUniqueId(this);
-                // Save innerHTML information in srcCode so we can access it later
-                // once we clean innerHTML (which is required since we don't want
-                // source code to be rendered on the screen)
-                this.srcCode = this.innerHTML;
-                const pySrc = await this.getPySrc();
-                this.innerHTML = '';
-
-                await app.plugins.beforePyScriptExec({ interpreter: interpreter, src: pySrc, pyScriptTag: this });
-                const result = (await pyExec(interpreter, pySrc, this)).result;
-                await app.plugins.afterPyScriptExec({
-                    interpreter: interpreter,
-                    src: pySrc,
-                    pyScriptTag: this,
-                    result: result,
-                });
-            } finally {
-                releaseLock();
-                app.decrementPendingTags();
-            }
+            // Save innerHTML information in srcCode so we can access it later
+            // once we clean innerHTML (which is required since we don't want
+            // source code to be rendered on the screen)
+            this.srcCode = this.innerHTML;
+            this.innerHTML = '';
+            await init(this, this._fetchSourceFallback);
         }
 
+        // TODO: is this useful at all? Can we move it away from the class?
         async getPySrc(): Promise<string> {
```

Comment:
I think we don't need this one anymore since your `fetchSource` method will do what `getPySrc` was doing 😄

Reply:
Done.